In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import us

from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.decomposition import PCA
from sklearn.datasets import fetch_olivetti_faces
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.ensemble import RandomForestRegressor

# Impact of Crime Rates and Air Quality on Housing Prices in the United States

## 1. Introduction

The real estate market is influenced by a wide range of factors, including economic conditions, demographic characteristics, and environmental quality. In recents years, there has been growing interest in understanding how non-traditional factors such as crime rates and air pollution impact housing prices.

This project aims to explore the relationship between housing prices and environmental conditions in the United States by combining multiple data sources, including housing data, crime statistics and air quality measurements. By integrating these datasets, the analysis provides a more comprehensive view of the factors that influence properly values.

In addition to traditionalhousing features such as income, population, and property characteristics, this study incorporates crime-related variables and air pollution indicators to evaluate their impact on real estate prices.

Machine learning techniques are applied to model these relathioships and assess the predictive power of environmental and socioeconomic variables. The results aim to hghlight whether environmental quality plays a significant role in housing valuation and how it compares to more conventional factors.

## 2. Data Overview

This project combines multiple datasets to analyze housing prices in relation to environmental and socioeconomic factors.

The main datasets used include:
* A housing dataset containg property-related features such as income, number of rooms, house age, price and population. => https://www.kaggle.com/datasets/vedavyasv/usa-housing?resource=download
* A crime dataset providing information aboout different types of crimes, including violent crime and property crime. => https://www.kaggle.com/datasets/kabhishm/united-states-crime-rates-by-city-population?select=crime_60_100.csv
* An air quality dataset containing poluution indicators such as NO2, O3, SO2 and CO levels. => https://www.kaggle.com/datasets/sogun3/uspollution

These dataset were merged based on geographical information (state-level data), allowing the integration of multiple independent data sources into a unified dataset.

The final dataset includes both numerical and categorical variables and represents multiple aspects of the living environment.

# 3. Data Processing

### 3.1 Housing Data

In [2]:
housing_data = pd.read_csv("data/USA_Housing.csv")

In [3]:
housing_data.head(10)

,Avg. Area Income,Avg. Area House Age,Avg. Area Number of Rooms,Avg. Area Number of Bedrooms,Area Population,Price,Address
0,79545.458574,5.682861,7.009188,4.09,23086.800503,1.059034e+06,"208 Michael Ferry Apt. 674\nLaurabury, NE 3701..."
1,79248.642455,6.002900,6.730821,3.09,40173.072174,1.505891e+06,"188 Johnson Views Suite 079\nLake Kathleen, CA..."
2,61287.067179,5.865890,8.512727,5.13,36882.159400,1.058988e+06,"9127 Elizabeth Stravenue\nDanieltown, WI 06482..."
3,63345.240046,7.188236,5.586729,3.26,34310.242831,1.260617e+06,USS Barnett\nFPO AP 44820
4,59982.197226,5.040555,7.839388,4.23,26354.109472,6.309435e+05,USNS Raymond\nFPO AE 09386
5,80175.754159,4.988408,6.104512,4.04,26748.428425,1.068138e+06,"06039 Jennifer Islands Apt. 443\nTracyport, KS..."
6,64698.463428,6.025336,8.147760,3.41,60828.249085,1.502056e+06,"4759 Daniel Shoals Suite 442\nNguyenburgh, CO ..."
7,78394.339278,6.989780,6.620478,2.42,36516.358972,1.573937e+06,"972 Joyce Viaduct\nLake William, TN 17778-6483"
8,59927.660813,5.362126,6.393121,2.30,29387.396003,7.988695e+05,USS Gilbert\nFPO AA 20957
9,81885.927184,4.423672,8.167688,6.10,40149.965749,1.545155e+06,Unit 9446 Box 0958\nDPO AE 97025


In [5]:
housing_data.shape

(5000, 7)

In [8]:
housing_data.isnull().sum()

Avg. Area Income                0
Avg. Area House Age             0
Avg. Area Number of Rooms       0
Avg. Area Number of Bedrooms    0
Area Population                 0
Price                           0
Address                         0
dtype: int64

In [4]:
housing_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Avg. Area Income              5000 non-null   float64
 1   Avg. Area House Age           5000 non-null   float64
 2   Avg. Area Number of Rooms     5000 non-null   float64
 3   Avg. Area Number of Bedrooms  5000 non-null   float64
 4   Area Population               5000 non-null   float64
 5   Price                         5000 non-null   float64
 6   Address                       5000 non-null   str    
dtypes: float64(6), str(1)
memory usage: 273.6 KB


In [7]:
housing_data.describe().T

,count,mean,std,min,25%,50%,75%,max
Avg. Area Income,5000.0,6.858311e+04,10657.991214,17796.631190,61480.562388,6.880429e+04,7.578334e+04,1.077017e+05
Avg. Area House Age,5000.0,5.977222e+00,0.991456,2.644304,5.322283,5.970429e+00,6.650808e+00,9.519088e+00
Avg. Area Number of Rooms,5000.0,6.987792e+00,1.005833,3.236194,6.299250,7.002902e+00,7.665871e+00,1.075959e+01
Avg. Area Number of Bedrooms,5000.0,3.981330e+00,1.234137,2.000000,3.140000,4.050000e+00,4.490000e+00,6.500000e+00
Area Population,5000.0,3.616352e+04,9925.650114,172.610686,29403.928702,3.619941e+04,4.286129e+04,6.962171e+04
Price,5000.0,1.232073e+06,353117.626581,15938.657923,997577.135049,1.232669e+06,1.471210e+06,2.469066e+06


The housing dataset contains 5000 observations and 7 variables providing key information about residentail properties.

The dataset includes several numerical features such as average area income, house age, number of rooms, number of bedrooms and population, as well as the target variable - housing price. Additionally, it contains and address column which represents the location of each property but is not directly used in the modeling process.

All numerical variables are complete, with no missing values, making the dataset suitable for direct analysis. This features are represented as floating-point numbers, while address columns is stored as string.

Summary statistics show that the average area income is approximately 68 000 while the average house price is around 1.23 million. The number of rooms average close to 7, indicating relatively large residential properties.

The dataset provides a solid foundation for analyzing the relationship between housing characteristics and property prices before integrating additional external factors such as crime and air quality.

In [10]:
housing_data["State"] = housing_data["Address"].str.extract(r'\b([A-Z]{2})\s+\d{5}\b')
housing_data = housing_data[housing_data.State.str.isupper()]

As part of the data processing stage, geographical information was extracted from the address field in the sousing dataset. SInce the original dataset did not include a separate state columns, a regular expression was included to identify and extract the state abbreviation from the address string. Specifically, the state code was extracted by identifying patterns consisting of two uppercase letters followed by ZIP code. This allowed the creation of a new **State** feature, which was later used to merge the housing dataset with external datasets containing crime statistics and air quality data. 

In [11]:
valid_states = [state.abbr for state in us.states.STATES]
housing_data = housing_data[housing_data.State.isin(valid_states)]
housing_data.isnull().sum()

Avg. Area Income                0
Avg. Area House Age             0
Avg. Area Number of Rooms       0
Avg. Area Number of Bedrooms    0
Area Population                 0
Price                           0
Address                         0
State                           0
dtype: int64

To ensure data quality only valid state abbreviations were retained. This step was essential for aligning the datasets and enabling accurate integration based on geographical information.